OpenFeature/src/main/java/dev/openfeature/sdk/async/Extensions.kt

```diff
@@ -1,33 +1,50 @@
 package dev.openfeature.sdk.async
 
+import dev.openfeature.sdk.FeatureProvider
+import dev.openfeature.sdk.OpenFeatureAPI
 import dev.openfeature.sdk.OpenFeatureClient
-import dev.openfeature.sdk.events.EventHandler
 import dev.openfeature.sdk.events.OpenFeatureEvents
 import dev.openfeature.sdk.events.observe
 import kotlinx.coroutines.CoroutineDispatcher
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.onStart
 import kotlinx.coroutines.flow.take
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.suspendCancellableCoroutine
 
-fun OpenFeatureClient.toAsync(dispatcher: CoroutineDispatcher = Dispatchers.IO): AsyncClient {
-    return AsyncClientImpl(this, dispatcher)
+fun OpenFeatureClient.toAsync(): AsyncClient? {
+    val provider = OpenFeatureAPI.getProvider()
+    return provider?.let {
+        AsyncClientImpl(
+            this,
+            it
+        )
+    }
 }
 
-internal fun observeProviderReady(
-    dispatcher: CoroutineDispatcher = Dispatchers.IO
-) = EventHandler.eventsObserver(dispatcher)
-    .observe<OpenFeatureEvents.ProviderReady>()
+internal fun FeatureProvider.observeProviderReady() = observe<OpenFeatureEvents.ProviderReady>()
     .onStart {
-        if (EventHandler.providerStatus().isProviderReady()) {
+        if (isProviderReady()) {
             this.emit(OpenFeatureEvents.ProviderReady)
         }
     }
 
-suspend fun awaitProviderReady(
+suspend fun OpenFeatureAPI.awaitProviderReady(
+    dispatcher: CoroutineDispatcher = Dispatchers.IO
+) {
+    val provider = getProvider()
+    requireNotNull(provider)
```

Comment:
Let's remove this too. I guess this function would just return right away if no provider is set?

Reply:
Hm... If no provider is set then `getProvider()` returns `null` and I assume `requireNotNull` will throw an exception. This is not great IMO.

Ideally this suspending function would just hang until a provider is set on `OpenFeatureAPI` and become ready. However I don't know how to solve that. Ideas?

Reply:
Is it completely crazy to setup a `flow` where we continously poll `getProvider()` until one becomes available?

Reply:
i thought about polling .. but isn't it too much maybe?

Reply:
we have 5 options : 
1- not have this function at all.
2- never return from this suspending if no provider set (bad because you won't know why you don't get out of the await function)
3- exception with correct message so you know you ask for this function early
4- poll until provider is set
5- return if provider is null (bad because returning implies the provider is ready which is not in this case)

IMO, we need to choose between option 4 or 3. I chose the option 3 but if that's violating the not throwing exception rule. I can try the polling thing.

Reply:
I may be missing some of the context here but would it be possible to return he no-op provider?

Reply:
@beeme1mr we updated the extension to be like following instead of waiting for provider in a different line and we are adding the `setProviderAndAwaitReady`.

```
suspend fun OpenFeatureAPI.setProviderAndAwaitReady(
    provider: FeatureProvider,
    dispatcher: CoroutineDispatcher
) {
    OpenFeatureAPI.setProvider(provider)
    provider.awaitReady(dispatcher)
}
```